azurerm/internal/services/iottimeseriesinsights/resource_iot_time_series_insights_access_policy.go

```diff
@@ -0,0 +1,212 @@
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringMatch(
+					regexp.MustCompile(`^[-\w\._\(\)]+$`),
+					"IoT Time Series Insights Access Policy name must contain only word characters, periods, underscores, hyphens, and parentheses.",
+				),
+			},
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"environment_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringMatch(
+					regexp.MustCompile(`^[-\w\._\(\)]+$`),
+					"IoT Time Series Insights Environment name must contain only word characters, periods, underscores, hyphens, and parentheses.",
+				),
+			},
+
+			"principal_object_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"description": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"roles": {
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+					ValidateFunc: validation.StringInSlice([]string{
+						string(timeseriesinsights.Contributor),
+						string(timeseriesinsights.Reader),
+					}, false),
+				},
+			},
+		},
+	}
+}
+
+func resourceArmIoTTimeSeriesInsightsAccessPolicyCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).IoTTimeSeriesInsights.AccessPoliciesClient
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	environmentName := d.Get("environment_name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	if d.IsNewResource() {
+		existing, err := client.Get(ctx, resourceGroup, environmentName, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for presence of existing IoT Time Series Insights Access Policy %q (Resource Group %q): %s", name, resourceGroup, err)
+			}
+		}
+
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_iot_time_series_insights_access_policy", *existing.ID)
+		}
+	}
+
+	policy := timeseriesinsights.AccessPolicyCreateOrUpdateParameters{
+		AccessPolicyResourceProperties: &timeseriesinsights.AccessPolicyResourceProperties{
+			Description:       utils.String(d.Get("description").(string)),
+			PrincipalObjectID: utils.String(d.Get("principal_object_id").(string)),
+			Roles:             expandIoTTimeSeriesInsightsAccessPolicyRoles(d.Get("roles").(*schema.Set).List()),
+		},
+	}
+
+	_, err := client.CreateOrUpdate(ctx, resourceGroup, environmentName, name, policy)
+	if err != nil {
```

Comment:
these lines cna be combines
```suggestion
	if _, err := client.CreateOrUpdate(ctx, resourceGroup, environmentName, name, policy); err != nil {
```